hpx/lcos/when_all.hpp

```diff
@@ -356,46 +357,43 @@ namespace hpx { namespace lcos
     template <typename Range>
     typename std::enable_if<traits::is_future_range<Range>::value,
         lcos::future<typename std::decay<Range>::type> >::type //-V659
-    when_all(Range&& values)
+    when_all(Range&& lazy_values)
     {
-        typedef detail::when_all_frame<util::tuple<Range> > frame_type;
+        typedef typename std::decay<Range>::type result_type;
+
+        result_type lazy_values_ = traits::acquire_future<result_type>()(lazy_values);
+
+        typedef detail::when_all_frame<
+                util::tuple<result_type>
+            > frame_type;
         typedef typename frame_type::init_no_addref init_no_addref;
 
         boost::intrusive_ptr<frame_type> p(new frame_type(
-            util::forward_as_tuple(std::move(values)), init_no_addref()),
+            util::forward_as_tuple(std::move(lazy_values_)), init_no_addref()),
```

Comment:
We don't move from the argument `lazy_values`, but from a local `lazy_values_` initialized by acquiring all the futures in the range given as argument. This is rather subtle, but a common pattern within the `when`s.

That said, it looks like the implementation of `acquire_future` might be insufficient to properly support non-owning ranges. I'll have to look into that.